File: autopilot_utils/src/autopilot_utils/trajectory_common.py
#!/usr/bin/env python3
import math
import random

from autopilot_msgs.msg import Trajectory, TrajectoryPoint
import rospy
from visualization_msgs.msg import Marker, MarkerArray
from geometry_msgs.msg import Pose, PoseStamped, Point
from nav_msgs.msg import Path
from autopilot_utils.pose_helper import *
import sys
# from nav_msgs.msg import Path


class TrajectoryManager:
    """
        This is a class for Trajectory related common operations
        Attributes:
            traj(autopilot_msgs/Trajectory)
        """

    def __init__(self, traj=None):
        """
               The constructor for TrajectoryManager class.

               Parameters:
                  traj(autopilot_msgs/Trajectory)
               """
        if traj:
            self._traj_in = traj
            self._traj_len = len(self._traj_in.points)
        else:
            self._traj_len = -1
            self._traj_in = None

    def update(self, traj):
        """
        updates the trajectory
        Parameters:
                  traj(autopilot_msgs/Trajectory)
        """
        self._traj_in = traj
        self._traj_len = len(self._traj_in.points)

    def get_len(self):
            return self._traj_len

    def get_traj_point(self, ind):
        # return 1
        if self._traj_len > 0 and ind < self._traj_len:
            return self._traj_in.points[ind]
        else:
            raise Exception("invalid index or out of bound")

    def next_point_within_dist(self, idx, dist_thr):
        # Find the next index which is dist_thr far from idx point.
        if self._traj_len > 0 and idx < self._traj_len:
            close_dis = self._traj_in.points[idx].accumulated_distance_m
            if abs(self._traj_in.points[idx].accumulated_distance_m - \
                   self._traj_in.points[-1].accumulated_distance_m) <= dist_thr:
                return self._traj_len - 1
            else:
                for i in range(idx, self._traj_len):
                    path_acc_distance = abs(self._traj_in.points[i].accumulated_distance_m - close_dis)
                    if path_acc_distance > dist_thr:
                        return i
        else:
            raise Exception("invalid index or out of bound")

    def find_closest_idx_with_dist_ang_thr(self, curr_pose, dist_thr, angle_thr):
        """
        Finds the closest point on trajectory within a distance threshold and angle threshold, returns true and
        index if one exists.
        Params:
            curr_pose : pose to which want to find close point
            dist_thr :
            angle_thr :
        Returns:
            Found, index (true/false, Int)
        """
        dist_min = sys.maxsize
        idx_min = -1
        curr_yaw = get_yaw(curr_pose.orientation)
        if self._traj_in is None:
            return False, idx_min
        for i in range(0, self._traj_len):
            traj_point = self._traj_in.points[i]
            dis = distance_btw_poses(curr_pose, traj_point.pose)
            if dis > dist_thr:
                continue
            yaw_diff = normalize_angle(curr_yaw - get_yaw(traj_point.pose.orientation))
            # print(yaw_diff)
            if abs(math.degrees(yaw_diff)) > angle_thr:
                continue
            if dis < dist_min:
                dist_min = dis
                idx_min = i
        if idx_min >= 0:
            var = True, idx_min
        else:
            var = False, idx_min
        return var

    def find_close_pose_after_index(self, curr_pose, prev_idx, search_distance=10):
        """
        This function returns the closest index, from prev_idx
        Params:
            curr_pose(pose) :
            prev_idx : search starts from this index
            search_distance : distance to search from prev_idx
        Returns:
            close_index: closest pose index to the curr_pose, from prev_idx.
        """

        # check for 10 meters from prev_idx
        dist_min = sys.maxsize
        idx_min = prev_idx
        if prev_idx >= self._traj_len - 1:
            return self._traj_len - 1

        for i in range(prev_idx, self._traj_len):
            if abs(self._traj_in.points[prev_idx].accumulated_distance_m -
                   self._traj_in.points[i].accumulated_distance_m) > search_distance:
                break
            traj_point = self._traj_in.points[i]
            dis = distance_btw_poses(curr_pose, traj_point.pose)
            if dis < dist_min:
                dist_min = dis
                idx_min = i
        return idx_min

    def to_path(self):
        """
        Returns path(Path) of Trajectory.
        """
        path_msg = Path()
        path_msg.header.frame_id = self._traj_len.header.frame_id
        for i in range(self._traj_len):
            pose_stamped = PoseStamped()
            pose_stamped.header.frame_id = self._traj_len.header.frame_id
            pose_stamped.pose = self._traj_in.points[i].pose
            path_msg.poses.append(pose_stamped)
        return path_msg

    def to_marker(self):
        """
        Returns marker(MarkerArray) of trajectory_points and  velocities.

        """
        marker_arr_msg = MarkerArray()
        marker_arr_msg.header.frame_id = self._traj_in.header.frame_id
        for i in range(self._traj_len):
            traj_point = self._traj_in.points[i]
            marker = Marker()
            marker.header.frame_id = self._traj_len.header.frame_id
            marker.type = marker.TEXT_VIEW_FACING
            marker.text = str(round(traj_point.longitudinal_velocity_mps, 2))
            marker.id = i
            marker.action = marker.ADD
            marker.scale.x = 0.1
            marker.scale.y = 0.1
            marker.scale.z = 0.1
            marker.color.a = 1.0
            marker.color.r = 0.0
            marker.color.g = 1.0
            marker.color.b = 0.0
            # marker.lifetime = rospy.Duration(duration)
            marker.pose = traj_point.pose
            marker_arr_msg.markers.append(marker)
        return marker_arr_msg

    def to_line(self, width):
        marker_arr = MarkerArray()
        marker = Marker()
        marker.header.frame_id = self._traj_in.header.frame_id
        marker.type = marker.LINE_STRIP
        # marker.text = str(round(traj_point.longitudinal_velocity_mps, 2))
        marker.id = 1
        marker.action = marker.ADD
        marker.scale.x = width
        marker.color.a = 1.0
        marker.color.r = 0.0
        marker.color.g = 1.0
        marker.color.b = 0.0

        # marker_arr_msg.header.frame_id = self._traj_in.header.frame_id
        for i in range(self._traj_len-10):
            pt = Point()
            pt.x = self._traj_in.points[i].pose.position.x
            pt.y = self._traj_in.points[i].pose.position.y
            pt.z = self._traj_in.points[i].pose.position.z
            marker.points.append(pt)
        marker_arr.markers.append(marker)
        marker = Marker()
        marker.header.frame_id = self._traj_in.header.frame_id
        marker.type = marker.LINE_STRIP
        # marker.text = str(round(traj_point.longitudinal_velocity_mps, 2))
        marker.id = 2
        marker.action = marker.ADD
        marker.scale.x = width
        marker.color.a = 1.0
        marker.color.r = 1.0
        marker.color.g = 0.0
        marker.color.b = 0.0

        # marker_arr_msg.header.frame_id = self._traj_in.header.frame_id
        for i in range(self._traj_len - 10, self._traj_len ):
            pt = Point()
            pt.x = self._traj_in.points[i].pose.position.x
            pt.y = self._traj_in.points[i].pose.position.y
            pt.z = self._traj_in.points[i].pose.position.z
            marker.points.append(pt)
        marker_arr.markers.append(marker)


        return marker_arr



if __name__ == "__main__":
    print("library for trajectory")
    rospy.init_node("test_trajectory_common")
    traj_pub = rospy.Publisher("/test_vis_mark", MarkerArray, queue_size=1, latch=True)
    raw = [2.5] * 20
    raw[0] = 1
    raw[-1] = 0
    traj_in = Trajectory()
    traj_in.header.frame_id = "map"
    f = 0
    for i in range(len(raw)):
        tp = TrajectoryPoint()
        tp.pose.position.x = i
        tp.pose.position.y = i
        tp.pose.orientation.w = 1
        tp.longitudinal_velocity_mps = raw[i]
        traj_in.points.append(tp)
    # print(traj_in.points[50])
    tm = TrajectoryManager(traj_in)
    #
    print(tm.to_line(0.5))
    traj_pub.publish(tm.to_line(0.5))
    rospy.loginfo("line published")
    rospy.spin()

File: autopilot_utils/src/trajectory_common.cpp
#include "autopilot_utils/trajectory_common.h"

namespace autopilot_utils{

TrajectoryHelper::TrajectoryHelper() {}

int TrajectoryHelper::getLength() { return traj.points.size(); }

autopilot_msgs::TrajectoryPoint
TrajectoryHelper::get_trajectory_point_by_index(int index) {
  /**
   * Returns the trajectory point at the given index.
   * Parameters:
   *     ind (int): The index of the trajectory point.
   * Returns:
   *     traj_point (autopilot_msgs/TrajectoryPoint): The trajectory point at
   * the given index. Raises: std::out_of_range: If the index is invalid or out
   * of bounds.
   */

  if (index >= 0 && index < traj.points.size())
    return traj.points[index];
  else {
    cerr << "Index out of range" << endl;
    return autopilot_msgs::TrajectoryPoint();
  }
}

double TrajectoryHelper::getCircumRadius(int ind, int index_lim = 10) {
  if ((ind - index_lim > 0) and (ind + index_lim < traj.points.size())) {
    vector<double> p1 = {traj.points[ind - index_lim].pose.position.x,
                         traj.points[ind - index_lim].pose.position.y};
    vector<double> p2 = {traj.points[ind].pose.position.x,
                         traj.points[ind].pose.position.y};
    vector<double> p3 = {traj.points[ind + index_lim].pose.position.x,
                         traj.points[ind + index_lim].pose.position.y};

    // return auto_nav::circumRadius(p1, p2, p3);
    //     den = 2 * ((x2 - x1) * (y3 - y2) - (y2 - y1) * (x3 - x2))
    double den = 2 * ((p2[0] - p1[0]) * (p3[1] - p2[1]) -
                      (p2[1] - p1[1]) * (p3[0] - p2[0]));
    if (den == 0) {
      cerr << "Failed: points are either collinear or not distinct" << endl;
      return 1000;
    }
    // float num = ((((x2 - x1) ** 2) + ((y2 - y1) ** 2)) * (((x3 - x2) ** 2) +
    // ((y3 - y2) ** 2)) * (((x1 - x3) ** 2) + ((y1 - y3) ** 2))) ** (0.5)
    double num = sqrt((pow(p2[0] - p1[0], 2) + pow(p2[1] - p1[1], 2)) *
                      (pow(p3[0] - p2[0], 2) + pow(p3[1] - p2[1], 2)) *
                      (pow(p1[0] - p3[0], 2) + pow(p1[1] - p3[1], 2)));

    double cirum_radius = num / den;
    return cirum_radius;
  }

  else {
    cerr << "Index out of bounds" << endl;
    return 0.0;
  }
}

void TrajectoryHelper::setTrajectory(const autopilot_msgs::Trajectory &trajectory) {
  traj = trajectory;
}

std::pair<bool, int32_t> TrajectoryHelper::find_closest_idx_with_dist_ang_thr(
    const geometry_msgs::Pose& robot_pose, double distance_th, double angle_th) {
  // Initialize variables to store the minimum squared distance and the
  // corresponding index
  double dist_squared_min = std::numeric_limits<double>::max();
  int32_t idx_min = -1;

  // Calculate the squared distance threshold
  double square_dist = distance_th * distance_th;

  // Iterate through each point in the trajectory
  for (int32_t i = 0; i < static_cast<int32_t>(traj.points.size()); ++i) {
    // Calculate the squared distance between the robot pose and the current
    // trajectory point
    const double ds =
        calcDistSquared2D(traj.points.at(i).pose.position, robot_pose.position);

    // Check if the squared distance exceeds the distance threshold
    if (ds > square_dist)
      continue; // Skip this point and move to the next one

    // Calculate the yaw angles of the robot pose and the trajectory point
    double yaw_pose = tf::getYaw(robot_pose.orientation);
    double yaw_ps = tf::getYaw(traj.points.at(i).pose.orientation);

    // Calculate the difference in yaw angles, normalizing the result
    double yaw_diff = normalizeAngle(yaw_pose - yaw_ps);

    // Check if the absolute value of the yaw difference exceeds the angle
    // threshold
    if (std::fabs(yaw_diff) > angle_th)
      continue; // Skip this point and move to the next one

    // Update the minimum squared distance and corresponding index if the
    // current point is closer
    if (ds < dist_squared_min) {
      dist_squared_min = ds;
      idx_min = i;
    }
  }

  // Return a pair indicating whether a valid index was found and the index
  // itself
  return (idx_min >= 0) ? std::make_pair(true, idx_min)
                        : std::make_pair(false, idx_min);
}

int TrajectoryHelper::find_close_pose_after_index(const geometry_msgs::Pose& curr_pose,
                                                  int prev_idx,
                                                  double search_distance) {
  /**
   * Returns the closest index from the previous index.
   * Parameters:
   *     curr_pose (geometry_msgs/Pose): The current pose.
   *     prev_idx (int): The index to start the search from.
   *     search_distance (double): The distance to search from the previous
   * index. Returns: close_index (int): The index of the closest pose to the
   * current pose from the previous index.
   */
  double dist_min = std::numeric_limits<double>::max();
  int idx_min = prev_idx;
  if (prev_idx >= traj.points.size()) {
    return traj.points.size() - 1;
  }
  for (int i = prev_idx; i < traj.points.size(); i++) {
    if (abs(traj.points[prev_idx].accumulated_distance_m -
            traj.points[i].accumulated_distance_m) > search_distance)
      break;

    double dist =
        sqrt(pow(curr_pose.position.x - traj.points[i].pose.position.x, 2) +
             pow(curr_pose.position.y - traj.points[i].pose.position.y, 2));
    if (dist < dist_min) {
      dist_min = dist;
      idx_min = i;
    }
  }
  return idx_min;
}

int TrajectoryHelper::next_point_within_dist(int idx, double dist_thr) {
  /**
   * Finds the next index which is dist_thr far from the given index.
   * Parameters:
   *     idx (int): The index to start the search from.
   *     dist_thr (double): The distance threshold.
   * Returns:
   *     next_idx (int): The index of the next point that is dist_thr far from
   * the given index. Raises: std::out_of_range: If the index is invalid or out
   * of bounds.
   */
  // Check if trajectory has points and if the provided index is within bounds
  if (traj.points.size() > 0 && idx < traj.points.size()) {
      // Get the accumulated distance of the closest point to the target distance
      double close_dis = traj.points[idx].accumulated_distance_m;
      
      // Check if the difference between the accumulated distance of the closest point
      // and the accumulated distance of the last point is within the threshold
      if (std::abs(traj.points[idx].accumulated_distance_m -
              traj.points.back().accumulated_distance_m) <= dist_thr)
        // Return the index of the last point if within threshold
        return traj.points.size() - 1;
      else {
        // Iterate through the trajectory points starting from the next point after the provided index
        for (int i = idx + 1; i < traj.points.size(); i++) {
          // Calculate the difference between the accumulated distance of the current point and the closest point
          double path_acc_distance =
              std::abs(traj.points[i].accumulated_distance_m - close_dis);
          // If the difference is greater than the threshold, return the index of the current point
          if (path_acc_distance > dist_thr)
            return i;
        }
      }
    }else
      // If trajectory has no points or provided index is out of bounds, return -1
      return -1;
}

nav_msgs::Path TrajectoryHelper::to_path() {
  /**
   * Returns the path of the trajectory.
   * Returns:
   *     path_msg (nav_msgs/Path): The path of the trajectory.
   */
  nav_msgs::Path path_msg;
  path_msg.header.frame_id = traj.header.frame_id;
  for (int i = 0; i < traj_len; i++) {
    geometry_msgs::PoseStamped pose_stamped;
    pose_stamped.header.frame_id = traj.header.frame_id;
    pose_stamped.pose = traj.points[i].pose;
    path_msg.poses.push_back(pose_stamped);
  }
  return path_msg;
}

} // End of namespace autopilot_utils
